### ISSUE TITLE
Support Unicode / UTF-8 characters

### ISSUE THREAD

**Issue by SamHasler** [org-trello/org-trello#307]
I have boards and cards with unicode characters in their names, and descriptions. After syncing they are not being converted into the correct characters in emacs. e.g.:

trello.org | org-trello characters | org-trello "as seen"
----------- | ------------ | --------
⌂ | â  | â\214\202
➊ | â | â\236\212

**Comment by ardumont** [comment]
Indeed, something is wrong.

With a card following this in trello:
```org
* IN-PROGRESS blahblah
:PROPERTIES:
:orgtrello-users: ardumont
:orgtrello-local-checksum: a3c236818b87c4d66761a9cd190991a805b62128ddd04215543576d424b346cf
:orgtrello-id: 567eb6cccb5c8a20bec078ab
:END:
  ah blah blah..
  - [-] Checklist
    - [ ] ⌂⌂ hey ho
```

Trying to sync from trello to org-trello buffer with such unicode character, the data is not retrieved correctly and the buffer contains some escaped characters (e.g '\342\214\202\342\214\202' here).

To reproduce:

```lisp
(let ((org-trello--user-logged-in "ardumont"))
  (orgtrello-controller-load-keys)
  (orgtrello-query-http-trello (orgtrello-api-get-full-card "567eb6cccb5c8a20bec078ab") 'sync))
#s(hash-table size 65 test equal rehash-size 1.5 rehash-threshold 0.8 data
              (:checklists
               (#s(hash-table size 65 test equal rehash-size 1.5 rehash-threshold 0.8 data
                              (:items
                               (#s(hash-table size 65 test equal rehash-size 1.5 rehash-threshold 0.8 data
                                              (:position 1501 :name "\342\214\202\342\214\202 hey ho" :id "56a8aa0f5082a04d2dadcb19" :checked "incomplete" :level 3)))
                               :position 1332 :card-id "567eb6cccb5c8a20bec078ab" :board-id "53fc6e11d9955d1d4b33aaab" :name "Checklist" :id "56a8aa0a49656b43bfc6d797" :level 2)))
               :comments nil :position 1114 :name "blahblah" :labels nil :member-ids
               ("4f2baa2f72b7c1293501cad3")
               :list-id "53fc6e17b928efc254b028fa" :due nil :desc "ah blah blah.." :closed nil :id "567eb6cccb5c8a20bec078ab" :level 1))
```

Even though, doing the same curl query does return the right content (as json):

```sh
curl https://api.trello.com/1/cards/567eb6cccb5c8a20bec078ab?fields=closed,dateLastActivity,desc,due,idChecklists,idList,idMembers,labels,name,pos&member_fields=fullName&actions=commentCard&action_fields=data&action_memberCreator_fields=username&key=<your-key>&token=<your-token> | tee blah.json
```

blah.json:
```json
{"id":"567eb6cccb5c8a20bec078ab","closed":false,"dateLastActivity":"2016-01-27T11:32:28.296Z","desc":"ah blah blah..","due":null,"idChecklists":["56a8aa0a49656b43bfc6d797"],"idList":"53fc6e17b928efc254b028fa","idMembers":["4f2baa2f72b7c1293501cad3"],"labels":[],"name":"blahblah","pos":1114,"actions":[{"id":"56a8a9fb73d1c2bedf10f9ba","data":{"list":{"name":"TODO","id":"53fc6e1754c24b2ab5d5eef5"},"board":{"shortLink":"RIlJAK05","name":"api test board","id":"53fc6e11d9955d1d4b33aaab"},"card":{"shortLink":"8xM2IEFo","idShort":870,"name":"blahbalh","id":"567eb6cccb5c8a20bec078ab"},"text":"⌂"},"memberCreator":{"id":"some-id","username":"ardumont"}}]}

```

Note:
The problem is not reproduced the other way around (sync to trello).

Multiple hints to investigate from top to bottom:
- orgtrello-query--http-parse
- json
- reques

**Comment by ardumont** [comment]
Insight about what I'm saying, here is the code that does parse the response from request https://github.com/org-trello/org-trello/blob/master/org-trello-query.el#L36-L39.

To follow on my last sentence, json-read seems ok:
```lisp
(require 'json)

(defun load-file (filepath)
  "Load the content of the FILEPATH."
  (with-temp-buffer
    (insert-file-contents (expand-file-name filepath))
    (buffer-substring-no-properties (point-min) (point-max))))

(let ((json-tryout-str (load-file "~/docs/org-trello-investigation/307/card.json")))
  (json-read-from-string json-tryout-str))
```

output:
```lisp
((actions .
          [((memberCreator
             (username . "ardumont")
             (id . "some-id"))
            (data
             (text . "⌂")
             (card
              (id . "567eb6cccb5c8a20bec078ab")
              (name . "blahbalh")
              (idShort . 870)
              (shortLink . "8xM2IEFo"))
             (board
              (id . "53fc6e11d9955d1d4b33aaab")
              (name . "api test board")
              (shortLink . "RIlJAK05"))
             (list
              (id . "53fc6e1754c24b2ab5d5eef5")
              (name . "TODO")))
            (id . "56a8a9fb73d1c2bedf10f9ba"))])
 (pos . 1114)
 (name . "blahblah")
 (labels .
   [])
 (idMembers .
            ["4f2baa2f72b7c1293501cad3"])
 (idList . "53fc6e17b928efc254b028fa")
 (idChecklists .
               ["56a8aa0a49656b43bfc6d797"])
 (due)
 (desc . "ah blah blah..")
 (dateLastActivity . "2016-01-27T11:32:28.296Z")
 (closed . :json-false)
 (id . "567eb6cccb5c8a20bec078ab"))
```
(card.json is a file containing the blah.json content from last comment)

Also, after that the parsing continues with orgtrello-query--http-parse which is also ok:

```lisp
(require 'dash)

(let ((json-tryout-str (load-file "~/docs/org-trello-investigation/307/card.json")))
  (-> json-tryout-str
      json-read-from-string
      orgtrello-data-parse-data))
#s(hash-table size 65 test equal rehash-size 1.5 rehash-threshold 0.8 data
              (:comments
               (#s(hash-table size 65 test equal rehash-size 1.5 rehash-threshold 0.8 data
                              (:comment-id "56a8a9fb73d1c2bedf10f9ba" :comment-text "⌂" :comment-date nil :comment-user "ardumont")))
               :position 1114 :name "blahblah" :labels nil :member-ids
               ("4f2baa2f72b7c1293501cad3")
               :list-id "53fc6e17b928efc254b028fa" :checklists
               ("56a8aa0a49656b43bfc6d797")
               :due nil :desc "ah blah blah.." :closed nil :id "567eb6cccb5c8a20bec078ab" :level 1))
```

So the problem may be before that (request?).

**Comment by archevel** [comment]
I spent some time debgging this. The problem did not seem to be with org-trello nor json-read. So I setup request.el to log messages at the debug level:
```
(custom-set-variables '(request-message-level 'debug))
```
  this makes it print the curl command it uses to fetch stuff and the response that it gets in the `*Messages*` buffer.
  I have created a single card named `New card with "Å" in the title` on a new board and the request.el printed out the response as:
```
REQUEST [debug] (buffer-string) =
HTTP/1.1 200 OK^M
Server: cloudflare-nginx^M
Date: Tue, 02 Feb 2016 05:20:31 GMT^M
Content-Type: application/json; charset=utf-8^M
Transfer-Encoding: chunked^M
Connection: keep-alive^M
Cache-Control: max-age=0, must-revalidate, no-cache, no-store^M
X-Content-Type-Options: nosniff^M
Strict-Transport-Security: max-age=15768000^M
X-XSS-Protection: 1; mode=block^M
X-FRAME-OPTIONS: DENY^M
X-Trello-Version: 1.576.0^M
X-Trello-Environment: Production^M
Access-Control-Allow-Origin: *^M
Access-Control-Allow-Methods: GET, PUT, POST, DELETE^M
X-Server-Time: 1454390431287^M
Expires: Thu, 01 Jan 1970 00:00:00^M
ETag: W/"f9-b25081fb"^M
Vary: Accept-Encoding^M
CF-RAY: 26e37282f7f53d0d-CPH^M
Content-Encoding: gzip^M
^M
[{"id":"56af71d0212b095266ec50dd","closed":false,"desc":"","due":null,"idBoard":"56af716a3498e989e40787f5","idList":"56af716ac35aef520491b9cb","idMembers":[],"labels":[],"name":"New card with \"\303\205\" in title","pos":431,"actions":[],"checklists":[]}]
```

So I tried the same curl command from the shell piping the response to a text file. Opening that file with `find-file` the Å character is displayed correctly. However opening the file with `find-file-literally` the Å shows up as `\303\205`. This leads me to believe that the problem is the default buffer encoding that causes problems. Digging around a bit further I discovered that patching line 979 in my request.el file by replacing:
```
(set-process-coding-system proc 'binary 'binary)
```
with:
```
(set-process-coding-system proc 'utf-8 'utf-8 )
```
and then updating the board the "Å" shows up as expected!

Not sure if this is an issue that should be raised with [https://github.com/abingham/emacs-request](https://github.com/abingham/emacs-request)? I have no idea about the wider consquences of this change (e.g. if downloading binaries breaks).

**Comment by ardumont** [comment]
Hello, wow thanks for continuing my started investigation ;).

This seems pretty interesting but i'm out of my depth here and without
means to look into it further up until next week (I intend to).

I think the binary exchange is what seems the more reasonable though. Then,
I suppose it's up to the client (org-trello through request here) to know
with what it is dealing with... Utf-8 here.

As a note, I did not know request changed maintainer, from @tkf to
@abingham.

Maybe asking for their help on the matter could unblock some pointers. @tkf
@abingham, do you have some ideas, pointers on the matter please?

Thanks again

Cheers,

**Comment by abingham** [comment]
I'll take look through this issue, but I can't make any promises. I took over maintenance because @tkf seems to have disappeared, and I had some important outstanding PRs for request.el. But I really don't know the guts of it much, and I'm not sure how much time I'll have to devote to it.

**Comment by abingham** [comment]
I think [this PR discussion in emacs-request](https://github.com/abingham/emacs-request/pull/5) might be apropos to this issue. @archevel showed that `set-process-coding-system` is implicated in the en/decoding problems, and that also shows up in the PR discussion. Based on the discussion, it looks like you (i.e. `org-trello`) might need to be using `decode-docing-string` differently. At least it's a clue!

Sorry I can't be any more insightful on this...too much other stuff on my plate :-/ But I'll do what I can to support the resolution of this issue.

**Comment by ardumont** [comment]
Thanks a bunch @abingham for the clue!
I'll look into it when i can.

Also thx for taking over request ;)

Cheers,

**Comment by archevel** [comment]
I managed to find a slightly better way to support unicode characters. In the [`orgtrello-data-parse-data`](https://github.com/org-trello/org-trello/blob/master/org-trello-data.el#L199)  I added the following special conversion for strings in the cond:
```
((stringp entities) (decode-coding-string entities 'utf-8))
```
(and removed the stringp from the in the check after it)

This seem to work and I'm fairly sure it is a better approach than changing the defaults in request.el. If I understand what is happening correctly this checks if the entities variable is a string and if it is reinterprets it as a string in utf-8. 

An even better approach would likely be to put this earlier in the parsing pipeline inside `orgtrello-query--http-parse`, but I couldn't figure out how to do that properly. Somehow the whole response should be passed through a function that decodes it as utf-8 and then passes it on to `json-read`. I would love to help out and do that since it seems the proper way to solve it, but my emacs lisp fu is severly limited :)

**Comment by ardumont** [comment]
> I managed to find a slightly better way to support unicode characters.
In the orgtrello-data-parse-data I added the following special conversion
for strings in the cond:

>```
>((stringp entities) (decode-coding-string >entities 'utf-8))
>```

> (and removed the stringp from the in the check after it)

This seems right.

Can you please show a full sample of code?

> This seem to work

Cool!

> And I'm fairly sure it is a better approach than changing the defaults in
request.el.

Well I was thinking along the lines of an org-trello hook to limit the
scope.

But yes.

> If I understand what is happening correctly this checks if the entities
variable is a string and if it is reinterprets it as a string in utf-8.

Yes, I think so too.

> An even better approach would likely be to put this earlier in the
parsing pipeline insideorgtrello-query--http-parse,

Indeed.

> but I couldn't figure out how to do that properly. Somehow the whole
response should be passed through a function that decodes it as utf-8 and
then passes it on to json-read.

Well, I'm also short on this one.
json-read does all the side-effecty magics...
It does not take any parameters...
I'd dug into the code... But I cannot yet...

> I would love to help out and do that since it seems the proper way to
solve it, but my emacs lisp fu is severly limited :)

Well you did manage to pull off some quite nice stuff up until now, bravo!

I have some ideas about 'defaround' or 'defadvice' to decorate a
function... (Since this one won't take any parameters and still do the
magic). But again I can't know for sure right now...

Cheers,

**Comment by archevel** [comment]
I managed to figure out how to change the encoding in the buffer that request.el uses by staring at the code long enough :) Made a PR with the fix!

**Comment by ardumont** [comment]
Cool. Thx.

On my way back to civilisation...
This seems ok from afar ;).

Can you please close the parenthesis on the previous lines as in other
function definitions?

Cheers,

**Comment by ardumont** [comment]
Thanks.
